avs_socket is unusable without a TLS backend enabled

`avs_socket.h` includes `avs_crypto_pki.h` which is not installed when all TLS backends are disabled.

Steps to reproduce:
```lang=sh
git clone https://github.com/AVSystem/avs_commons
mkdir avs_commons/build
cd avs_commons/build

cmake -DCMAKE_INSTALL_PREFIX=$PWD/install ..
make install
gcc -I $PWD/install/include -x c - <<<"#include <avsystem/commons/avs_socket.h>"   
# compiles fine, linker complains about missing main 

rm -rf ./*

cmake -DWITH_OPENSSL=OFF -DWITH_MBEDTLS=OFF -DWITH_TINYDTLS=OFF -DCMAKE_INSTALL_PREFIX=$PWD/install ..
make install
gcc -I $PWD/install/include -x c - <<<"#include <avsystem/commons/avs_socket.h>"   
# In file included from <stdin>:1:
# /home/dex/projects/avs_commons/build/install/include/avsystem/commons/avs_socket.h:30:10: fatal error: avsystem/commons/avs_crypto_pki.h: No such file or directory
#    30 | #include <avsystem/commons/avs_crypto_pki.h>
#       |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
# compilation terminated.
```

--- a/include_public/avsystem/commons/avs_commons_config.h.in
+++ b/include_public/avsystem/commons/avs_commons_config.h.in
@@ -18,7 +18,7 @@
 #define AVS_COMMONS_CONFIG_H
 
 /**
- * @file avs_avs_commons_config.h
+ * @file avs_commons_config.h
  *
  * avs_commons library configuration.
  *

--- a/include_public/avsystem/commons/avs_http.h
+++ b/include_public/avsystem/commons/avs_http.h
@@ -196,6 +196,7 @@ avs_http_t *avs_http_new(const avs_http_buffer_sizes_t *buffer_sizes);
  */
 void avs_http_free(avs_http_t *http);
 
+#ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 /**
  * Sets SSL/TLS configuration for sockets created by the the HTTP client when
  * connecting to HTTPS addresses.
@@ -211,6 +212,7 @@ void avs_http_free(avs_http_t *http);
 void avs_http_ssl_configuration(
         avs_http_t *http,
         const volatile avs_net_ssl_configuration_t *ssl_configuration);
+#endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 /**
  * Callback function type used by @ref avs_http_ssl_pre_connect_cb .

--- a/include_public/avsystem/commons/avs_socket.h
+++ b/include_public/avsystem/commons/avs_socket.h
@@ -27,12 +27,15 @@
 #include <stdlib.h>
 
 #include <avsystem/commons/avs_commons_config.h>
-#include <avsystem/commons/avs_crypto_pki.h>
 #include <avsystem/commons/avs_defs.h>
 #include <avsystem/commons/avs_errno.h>
 #include <avsystem/commons/avs_memory.h>
-#include <avsystem/commons/avs_prng.h>
 #include <avsystem/commons/avs_time.h>
+
+#ifdef AVS_COMMONS_WITH_AVS_CRYPTO
+#    include <avsystem/commons/avs_crypto_pki.h>
+#    include <avsystem/commons/avs_prng.h>
+#endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 #ifdef AVS_COMMONS_NET_WITH_X509
 #    warning \
@@ -85,20 +88,6 @@ typedef struct avs_net_socket_struct avs_net_socket_t;
  * interface. For POSIX interfaces it is array of IF_NAMESIZE characters.
  */
 typedef char avs_net_socket_interface_name_t[IF_NAMESIZE];
-
-/**
- * Function type for callbacks to be executed for additional SSL configuration.
- *
- * Note that the @ref library_ssl_context parameter is a pointer to a native
- * SSL context object of the SSL library in use. It shall be cast to
- * <c>SSL_CTX *</c> for OpenSSL or <c>ssl_context *</c> for XySSL-derivatives.
- *
- * @param library_ssl_context pointer to a native SSL context object of the
- *                            SSL library in use
- *
- * @return 0 on success, negative value on failure
- */
-typedef int avs_ssl_additional_configuration_clb_t(void *library_ssl_context);
 
 /**
  * Structure that contains additional configuration options for creating TCP and
@@ -217,6 +206,21 @@ typedef struct {
     avs_net_af_t preferred_family;
 } avs_net_socket_configuration_t;
 
+#ifdef AVS_COMMONS_WITH_AVS_CRYPTO
+/**
+ * Function type for callbacks to be executed for additional SSL configuration.
+ *
+ * Note that the @ref library_ssl_context parameter is a pointer to a native
+ * SSL context object of the SSL library in use. It shall be cast to
+ * <c>SSL_CTX *</c> for OpenSSL or <c>ssl_context *</c> for XySSL-derivatives.
+ *
+ * @param library_ssl_context pointer to a native SSL context object of the
+ *                            SSL library in use
+ *
+ * @return 0 on success, negative value on failure
+ */
+typedef int avs_ssl_additional_configuration_clb_t(void *library_ssl_context);
+
 /**
  * Available SSL versions that can be used by SSL sockets.
  */
@@ -333,6 +337,7 @@ typedef struct {
     avs_time_duration_t min;
     avs_time_duration_t max;
 } avs_net_dtls_handshake_timeouts_t;
+#endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 /**
  * Category for @ref avs_error_t containing a (D)TLS Alert.
@@ -369,6 +374,7 @@ static inline uint8_t avs_net_ssl_alert_description(avs_error_t error) {
     return (uint8_t) error.code;
 }
 
+#ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 typedef struct {
     /** Array of ciphersuite IDs, or NULL to enable all ciphers */
     uint32_t *ids;
@@ -480,6 +486,7 @@ typedef struct {
      */
     avs_crypto_prng_ctx_t *prng_ctx;
 } avs_net_ssl_configuration_t;
+#endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 typedef enum {
     /**
@@ -674,6 +681,7 @@ avs_error_t
 avs_net_tcp_socket_create(avs_net_socket_t **socket,
                           const avs_net_socket_configuration_t *config);
 
+#ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 avs_error_t
 avs_net_dtls_socket_create(avs_net_socket_t **socket,
                            const avs_net_ssl_configuration_t *config);
@@ -681,7 +689,7 @@ avs_net_dtls_socket_create(avs_net_socket_t **socket,
 avs_error_t
 avs_net_ssl_socket_create(avs_net_socket_t **socket,
                           const avs_net_ssl_configuration_t *config);
-
+#endif // AVS_COMMONS_WITH_AVS_CRYPTO
 /**@}*/
 
 /**
@@ -739,6 +747,7 @@ avs_error_t avs_net_socket_connect(avs_net_socket_t *socket,
 avs_error_t avs_net_socket_decorate(avs_net_socket_t *socket,
                                     avs_net_socket_t *backend_socket);
 
+#ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 /**
  * @name Sockets decorators
  *
@@ -768,6 +777,7 @@ avs_net_ssl_socket_decorate_in_place(avs_net_socket_t **socket,
                                      const avs_net_ssl_configuration_t *config);
 
 /**@}*/
+#endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 /**
  * Sends exactly @p buffer_length bytes from @p buffer to @p socket.

--- a/src/avs_commons_init.h
+++ b/src/avs_commons_init.h
@@ -34,7 +34,8 @@
 #endif
 
 #if !(defined(AVS_COMMONS_WITH_OPENSSL) || defined(AVS_COMMONS_WITH_MBEDTLS) \
-      || defined(AVS_COMMONS_WITH_TINYDTLS))
+      || defined(AVS_COMMONS_WITH_TINYDTLS))                                 \
+        && !defined(WITHOUT_SSL)
 #    define WITHOUT_SSL
 #endif
 

--- a/src/crypto/CMakeLists.txt
+++ b/src/crypto/CMakeLists.txt
@@ -13,6 +13,7 @@
 # limitations under the License.
 
 set(AVS_CRYPTO_PUBLIC_HEADERS
+    "${AVS_COMMONS_SOURCE_DIR}/include_public/avsystem/commons/avs_crypto_pki.h"
     "${AVS_COMMONS_SOURCE_DIR}/include_public/avsystem/commons/avs_prng.h")
 
 set(AVS_CRYPTO_COMMON_SOURCES
@@ -53,12 +54,6 @@ target_link_libraries(avs_crypto_core INTERFACE avs_commons_global_headers avs_u
 
 if(WITH_INTERNAL_LOGS)
     target_link_libraries(avs_crypto_core INTERFACE avs_log)
-endif()
-
-if(WITH_OPENSSL OR WITH_MBEDTLS)
-    set(AVS_CRYPTO_PUBLIC_HEADERS
-        ${AVS_CRYPTO_PUBLIC_HEADERS}
-        "${AVS_COMMONS_SOURCE_DIR}/include_public/avsystem/commons/avs_crypto_pki.h")
 endif()
 
 if(WITH_OPENSSL)

--- a/src/crypto/avs_crypto_utils.c
+++ b/src/crypto/avs_crypto_utils.c
@@ -22,6 +22,7 @@
 #    include <string.h>
 
 #    include <avsystem/commons/avs_crypto_pki.h>
+#    include <avsystem/commons/avs_memory.h>
 
 #    include "avs_crypto_utils.h"
 

--- a/src/http/CMakeLists.txt
+++ b/src/http/CMakeLists.txt
@@ -44,7 +44,7 @@ add_library(avs_http STATIC
             avs_http_stream.c
             avs_stream_methods.c)
 
-target_link_libraries(avs_http PUBLIC avs_commons_global_headers avs_algorithm avs_net avs_stream avs_stream_md5 avs_stream_net avs_utils avs_list)
+target_link_libraries(avs_http PUBLIC avs_commons_global_headers avs_algorithm avs_net_core avs_stream avs_stream_md5 avs_stream_net avs_utils avs_list)
 
 if(WITH_AVS_HTTP_ZLIB)
     avs_find_library("find_package(ZLIB REQUIRED)")
@@ -57,7 +57,7 @@ install(FILES ${AVS_HTTP_PUBLIC_HEADERS}
         DESTINATION ${INCLUDE_INSTALL_DIR}/avsystem/commons)
 
 avs_add_test(NAME avs_http
-             LIBS avs_http
+             LIBS avs_http avs_net
              SOURCES
              $<TARGET_PROPERTY:avs_http,SOURCES>
              ${AVS_COMMONS_SOURCE_DIR}/tests/http/test_close.c

--- a/src/http/avs_client.c
+++ b/src/http/avs_client.c
@@ -61,11 +61,13 @@ void avs_http_free(avs_http_t *http) {
     }
 }
 
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 void avs_http_ssl_configuration(
         avs_http_t *http,
         const volatile avs_net_ssl_configuration_t *ssl_configuration) {
     http->ssl_configuration = ssl_configuration;
 }
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 void avs_http_ssl_pre_connect_cb(avs_http_t *http,
                                  avs_http_ssl_pre_connect_cb_t *cb,

--- a/src/http/avs_client.h
+++ b/src/http/avs_client.h
@@ -38,7 +38,9 @@ struct avs_http {
     avs_http_ssl_pre_connect_cb_t *ssl_pre_connect_cb;
     void *ssl_pre_connect_cb_arg;
 
+#ifdef AVS_COMMONS_WITH_AVS_CRYPTO
     const volatile avs_net_ssl_configuration_t *ssl_configuration;
+#endif // AVS_COMMONS_WITH_AVS_CRYPTO
     const volatile avs_net_socket_configuration_t *tcp_configuration;
 };
 

--- a/src/http/avs_http_stream.c
+++ b/src/http/avs_http_stream.c
@@ -89,10 +89,15 @@ static const char *resolve_port(const avs_url_t *parsed_url) {
 avs_error_t _avs_http_socket_new(avs_net_socket_t **out,
                                  avs_http_t *client,
                                  const avs_url_t *url) {
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
     avs_net_ssl_configuration_t ssl_config_full;
+#    else  // AVS_COMMONS_WITH_AVS_CRYPTO
+    avs_net_socket_configuration_t tcp_config_full;
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
     LOG(TRACE, _("http_new_socket"));
     assert(out != NULL);
     *out = NULL;
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
     if (client->ssl_configuration) {
         ssl_config_full = *client->ssl_configuration;
     } else {
@@ -101,6 +106,13 @@ avs_error_t _avs_http_socket_new(avs_net_socket_t **out,
     if (client->tcp_configuration) {
         ssl_config_full.backend_configuration = *client->tcp_configuration;
     }
+#    else  // AVS_COMMONS_WITH_AVS_CRYPTO
+    if (client->tcp_configuration) {
+        tcp_config_full = *client->tcp_configuration;
+    } else {
+        memset(&tcp_config_full, 0, sizeof(tcp_config_full));
+    }
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
     const char *host = avs_url_host(url);
     const char *port = resolve_port(url);
     avs_error_t err = avs_errno(AVS_EINVAL);
@@ -108,9 +120,15 @@ avs_error_t _avs_http_socket_new(avs_net_socket_t **out,
     case HTTP_URI_PROTOCOL_HTTP:
         LOG(TRACE, _("creating TCP socket"));
         err = avs_net_tcp_socket_create(out,
-                                        &ssl_config_full.backend_configuration);
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
+                                        &ssl_config_full.backend_configuration
+#    else  // AVS_COMMONS_WITH_AVS_CRYPTO
+                                        &tcp_config_full
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
+        );
         break;
     case HTTP_URI_PROTOCOL_HTTPS:
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
         LOG(TRACE, _("creating SSL socket"));
         if (avs_is_ok((err = avs_net_ssl_socket_create(out, &ssl_config_full)))
                 && client->ssl_pre_connect_cb
@@ -120,6 +138,7 @@ avs_error_t _avs_http_socket_new(avs_net_socket_t **out,
             avs_net_socket_cleanup(out);
         }
         break;
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
     case HTTP_URI_PROTOCOL_UNKNOWN:
         break;
     }

--- a/src/net/avs_api.c
+++ b/src/net/avs_api.c
@@ -37,6 +37,7 @@ VISIBILITY_SOURCE_BEGIN
 
 const avs_time_duration_t AVS_NET_SOCKET_DEFAULT_RECV_TIMEOUT = { 30, 0 };
 
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 avs_net_security_info_t avs_net_security_info_from_psk(avs_net_psk_info_t psk) {
     avs_net_security_info_t result;
     memset(&result, 0, sizeof(result));
@@ -53,6 +54,7 @@ avs_net_security_info_from_certificates(avs_net_certificate_info_t info) {
     result.data.cert = info;
     return result;
 }
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 #    ifdef AVS_COMMONS_NET_WITH_SOCKET_LOG
 static int _avs_net_socket_debug = 0;
@@ -283,6 +285,7 @@ static avs_error_t create_bare_socket(avs_net_socket_t **socket,
     return socket_constructor(socket, configuration);
 }
 
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 static avs_error_t decorate_socket_in_place(
         avs_net_socket_t **socket,
         avs_error_t (*new_socket_constructor)(
@@ -311,6 +314,7 @@ avs_error_t avs_net_ssl_socket_decorate_in_place(
         avs_net_socket_t **socket, const avs_net_ssl_configuration_t *config) {
     return decorate_socket_in_place(socket, avs_net_ssl_socket_create, config);
 }
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 #    ifdef AVS_COMMONS_NET_WITH_SOCKET_LOG
 
@@ -675,10 +679,11 @@ avs_net_tcp_socket_create(avs_net_socket_t **socket,
     return init_debug_socket_if_applicable(socket, err);
 }
 
+#    ifdef AVS_COMMONS_WITH_AVS_CRYPTO
 avs_error_t
 avs_net_dtls_socket_create(avs_net_socket_t **socket,
                            const avs_net_ssl_configuration_t *config) {
-#    ifndef WITHOUT_SSL
+#        ifndef WITHOUT_SSL
     if (!config->prng_ctx) {
         LOG(ERROR, _("PRNG ctx MUST NOT be NULL"));
         return avs_errno(AVS_EINVAL);
@@ -686,18 +691,18 @@ avs_net_dtls_socket_create(avs_net_socket_t **socket,
     avs_error_t err =
             create_bare_socket(socket, _avs_net_create_dtls_socket, config);
     return init_debug_socket_if_applicable(socket, err);
-#    else  // WITHOUT_SSL
+#        else  // WITHOUT_SSL
     (void) socket;
     (void) config;
     LOG(ERROR, _("could not create secure socket: (D)TLS support is disabled"));
     return avs_errno(AVS_ENOTSUP);
-#    endif // WITHOUT_SSL
+#        endif // WITHOUT_SSL
 }
 
 avs_error_t
 avs_net_ssl_socket_create(avs_net_socket_t **socket,
                           const avs_net_ssl_configuration_t *config) {
-#    ifndef WITHOUT_SSL
+#        ifndef WITHOUT_SSL
     if (!config->prng_ctx) {
         LOG(ERROR, _("PRNG ctx MUST NOT be NULL"));
         return avs_errno(AVS_EINVAL);
@@ -705,12 +710,13 @@ avs_net_ssl_socket_create(avs_net_socket_t **socket,
     avs_error_t err =
             create_bare_socket(socket, _avs_net_create_ssl_socket, config);
     return init_debug_socket_if_applicable(socket, err);
-#    else  // WITHOUT_SSL
+#        else  // WITHOUT_SSL
     (void) socket;
     (void) config;
     LOG(ERROR, _("could not create secure socket: (D)TLS support is disabled"));
     return avs_errno(AVS_ENOTSUP);
-#    endif // WITHOUT_SSL
+#        endif // WITHOUT_SSL
 }
+#    endif // AVS_COMMONS_WITH_AVS_CRYPTO
 
 #endif // AVS_COMMONS_WITH_AVS_NET

--- a/src/stream/net/CMakeLists.txt
+++ b/src/stream/net/CMakeLists.txt
@@ -26,7 +26,7 @@ add_library(avs_stream_net STATIC
             avs_stream_net.c
             avs_netbuf.c)
 
-target_link_libraries(avs_stream_net PUBLIC avs_stream avs_buffer avs_net)
+target_link_libraries(avs_stream_net PUBLIC avs_stream avs_buffer avs_net_core)
 
 avs_install_export(avs_stream_net stream)
 install(FILES ${AVS_STREAM_NET_PUBLIC_HEADERS}

--- a/tests/net/socket_common.h
+++ b/tests/net/socket_common.h
@@ -102,6 +102,8 @@ run_socket_set_opt_test_cases(avs_net_socket_t *socket,
         case AVS_NET_SOCKET_OPT_BYTES_RECEIVED:
             opt_val.bytes_received = 321;
             break;
+        case AVS_NET_SOCKET_OPT_DANE_TLSA_ARRAY:
+            AVS_UNREACHABLE("unsupported case");
         }
 
         if (test_cases[i].expected_result == SUCCESS) {

--- a/tools/dtls_echo_server.c
+++ b/tools/dtls_echo_server.c
@@ -222,7 +222,7 @@ generate_cookie(SSL *ssl, unsigned char *cookie, unsigned int *cookie_len) {
 }
 
 static int
-verify_cookie(SSL *ssl, unsigned char *cookie, unsigned int cookie_len) {
+verify_cookie(SSL *ssl, const unsigned char *cookie, unsigned int cookie_len) {
     unsigned char *buffer, result[EVP_MAX_MD_SIZE];
     size_t length = 0;
     unsigned resultlength;
@@ -290,11 +290,8 @@ verify_cookie(SSL *ssl, unsigned char *cookie, unsigned int cookie_len) {
 }
 
 struct pass_info {
-    union {
-        struct sockaddr_storage ss;
-        struct sockaddr_in6 s6;
-        struct sockaddr_in s4;
-    } server_addr, client_addr;
+    const BIO_ADDR *server_addr;
+    const BIO_ADDR *client_addr;
     SSL *ssl;
     int fd;
 };
@@ -312,30 +309,17 @@ static int dtls_verify_callback(int ok, X509_STORE_CTX *ctx) {
 static void connection_handle(struct pass_info *pinfo) {
     int len;
     char buf[BUFFER_SIZE];
-    char addrbuf[INET6_ADDRSTRLEN];
     SSL *ssl = pinfo->ssl;
     int reading = 0, ret;
     int num_timeouts = 0, max_timeouts = 5;
 
-    OPENSSL_assert(pinfo->client_addr.ss.ss_family
-                   == pinfo->server_addr.ss.ss_family);
-    switch (pinfo->client_addr.ss.ss_family) {
-    case AF_INET:
-        connect(pinfo->fd, (struct sockaddr *) &pinfo->client_addr,
-                sizeof(struct sockaddr_in));
-        break;
-    case AF_INET6:
-        connect(pinfo->fd, (struct sockaddr *) &pinfo->client_addr,
-                sizeof(struct sockaddr_in6));
-        break;
-    default:
-        OPENSSL_assert(0);
-        break;
-    }
+    OPENSSL_assert(BIO_ADDR_family(pinfo->client_addr)
+                   == BIO_ADDR_family(pinfo->server_addr));
+    BIO_connect(pinfo->fd, pinfo->client_addr, 0);
 
     /* Set new fd and set BIO to connected */
     BIO_ctrl(SSL_get_rbio(ssl), BIO_CTRL_DGRAM_SET_CONNECTED, 0,
-             &pinfo->client_addr.ss);
+             (void *) (intptr_t) pinfo->client_addr);
 
     /* Finish handshake */
     do {
@@ -348,17 +332,11 @@ static void connection_handle(struct pass_info *pinfo) {
     }
 
     if (verbose) {
-        if (pinfo->client_addr.ss.ss_family == AF_INET) {
-            printf("\naccepted connection from %s:%d\n",
-                   inet_ntop(AF_INET, &pinfo->client_addr.s4.sin_addr, addrbuf,
-                             INET6_ADDRSTRLEN),
-                   ntohs(pinfo->client_addr.s4.sin_port));
-        } else {
-            printf("\naccepted connection from %s:%d\n",
-                   inet_ntop(AF_INET6, &pinfo->client_addr.s6.sin6_addr,
-                             addrbuf, INET6_ADDRSTRLEN),
-                   ntohs(pinfo->client_addr.s6.sin6_port));
-        }
+        char *hostname = BIO_ADDR_hostname_string(pinfo->client_addr, 0);
+        char *port = BIO_ADDR_service_string(pinfo->client_addr, 1);
+        printf("\naccepted connection from %s:%s\n", hostname, port);
+        OPENSSL_free(hostname);
+        OPENSSL_free(port);
     }
 
     if (veryverbose && SSL_get_peer_certificate(ssl)) {
@@ -463,53 +441,28 @@ cleanup:
     close(pinfo->fd);
 #endif
     SSL_free(ssl);
-    ERR_remove_state(0);
     if (verbose)
         printf("done, connection closed.\n");
 }
 
-static void start_server(uint16_t port,
-                         char *local_address,
+static void start_server(const char *port,
+                         const char *local_address,
                          const char *ca_file,
                          const char *pkey_file) {
     int fd;
-    union {
-        struct sockaddr_storage ss;
-        struct sockaddr_in s4;
-        struct sockaddr_in6 s6;
-    } server_addr, client_addr;
     SSL_CTX *ctx;
     SSL *ssl;
     BIO *bio;
     struct timeval timeout;
     struct pass_info info;
-    const int on = 1, off = 0;
+    const int on = 1;
 
-    memset(&server_addr, 0, sizeof(struct sockaddr_storage));
-    if (strlen(local_address) == 0) {
-        server_addr.s6.sin6_family = AF_INET6;
-#ifdef HAVE_SIN6_LEN
-        server_addr.s6.sin6_len = sizeof(struct sockaddr_in6);
-#endif
-        server_addr.s6.sin6_addr = in6addr_any;
-        server_addr.s6.sin6_port = htons(port);
-    } else {
-        if (inet_pton(AF_INET, local_address, &server_addr.s4.sin_addr) == 1) {
-            server_addr.s4.sin_family = AF_INET;
-#ifdef HAVE_SIN_LEN
-            server_addr.s4.sin_len = sizeof(struct sockaddr_in);
-#endif
-            server_addr.s4.sin_port = htons(port);
-        } else if (inet_pton(AF_INET6, local_address, &server_addr.s6.sin6_addr)
-                   == 1) {
-            server_addr.s6.sin6_family = AF_INET6;
-#ifdef HAVE_SIN6_LEN
-            server_addr.s6.sin6_len = sizeof(struct sockaddr_in6);
-#endif
-            server_addr.s6.sin6_port = htons(port);
-        } else {
-            return;
-        }
+    BIO_ADDRINFO *server_addrinfo = NULL;
+    BIO_lookup((local_address && *local_address) ? local_address : NULL,
+               (port && *port) ? port : NULL, BIO_LOOKUP_SERVER, AF_UNSPEC,
+               SOCK_DGRAM, &server_addrinfo);
+    if (!server_addrinfo) {
+        return;
     }
 
     OpenSSL_add_ssl_algorithms();
@@ -546,7 +499,8 @@ static void start_server(uint16_t port,
     WSAStartup(MAKEWORD(2, 2), &wsaData);
 #endif
 
-    fd = socket(server_addr.ss.ss_family, SOCK_DGRAM, 0);
+    fd = socket(BIO_ADDR_family(BIO_ADDRINFO_address(server_addrinfo)),
+                SOCK_DGRAM, 0);
     if (fd < 0) {
         perror("socket");
         exit(-1);
@@ -564,19 +518,15 @@ static void start_server(uint16_t port,
 #    endif
 #endif
 
-    if (server_addr.ss.ss_family == AF_INET) {
-        bind(fd, (const struct sockaddr *) &server_addr,
-             sizeof(struct sockaddr_in));
-    } else {
-        setsockopt(fd, IPPROTO_IPV6, IPV6_V6ONLY, (const char *) &off,
-                   sizeof(off));
-        bind(fd, (const struct sockaddr *) &server_addr,
-             sizeof(struct sockaddr_in6));
-    }
+    BIO_bind(fd, BIO_ADDRINFO_address(server_addrinfo), 0);
 
     kill(getppid(), SIGUSR1);
 
-    memset(&client_addr, 0, sizeof(struct sockaddr_storage));
+    BIO_ADDR *client_addr = BIO_ADDR_new();
+    if (!client_addr) {
+        perror("out of memory");
+        exit(-1);
+    }
 
     /* Create BIO */
     bio = BIO_new_dgram(fd, BIO_NOCLOSE);
@@ -592,15 +542,18 @@ static void start_server(uint16_t port,
     SSL_set_options(ssl, SSL_OP_COOKIE_EXCHANGE);
     SSL_set_accept_state(ssl);
 
-    while (DTLSv1_listen(ssl, &client_addr) <= 0)
+    while (DTLSv1_listen(ssl, client_addr) <= 0)
         ;
 
-    memcpy(&info.server_addr, &server_addr, sizeof(struct sockaddr_storage));
-    memcpy(&info.client_addr, &client_addr, sizeof(struct sockaddr_storage));
+    info.server_addr = BIO_ADDRINFO_address(server_addrinfo);
+    info.client_addr = client_addr;
     info.ssl = ssl;
     info.fd = fd;
 
     connection_handle(&info);
+
+    BIO_ADDRINFO_free(server_addrinfo);
+    BIO_ADDR_free(client_addr);
 
 #ifdef WIN32
     WSACleanup();
@@ -608,7 +561,7 @@ static void start_server(uint16_t port,
 }
 
 int main(int argc, char **argv) {
-    uint16_t port = 23232;
+    const char *port = "23232";
     int length = 100;
     char local_addr[INET6_ADDRSTRLEN + 1];
 
@@ -635,7 +588,7 @@ int main(int argc, char **argv) {
                 goto cmd_err;
             int port_int = atoi(*++argv);
             if (0 <= port_int && port_int < UINT16_MAX) {
-                port = (uint16_t) port_int;
+                port = *argv;
             } else {
                 goto cmd_err;
             }

